cirq-google/cirq_google/ops/gateset.py

```diff
@@ -0,0 +1,112 @@
+from typing import cast, Optional, Union, Type, List, Sequence
+
+import cirq
+from cirq_google.ops import fsim_gate_family, sycamore_gate
+from cirq.protocols.decompose_protocol import DecomposeResult
+import cirq_google.transformers.target_gatesets as cg_target_gatesets
+
+
+def _build_equivalence_gate_families(device_gateset: cirq.Gateset) -> List[cirq.GateFamily]:
+    fsim_gates = [
+        cast(fsim_gate_family.POSSIBLE_FSIM_GATES, g)
+        for g in device_gateset.gates
+        if isinstance(g, tuple(fsim_gate_family.POSSIBLE_FSIM_GATE_TYPES))
+    ]
+    return [fsim_gate_family.FSimGateFamily(gates_to_accept=fsim_gates)]
+
+    # TODO(verult) PhasedXZGateFamily
+
+
+class Gateset(cirq.CompilationTargetGateset):
+    """Represents the gateset of a Google device.
+
+    This gateset accept circuits that use any Cirq gate representation of valid device gates.
+
+    It can also be used as an argument in cirq.optimize_for_target_gateset() to transform your
+    circuit into one that consists of valid gates for the target device. In cases where multiple
+    `CompilationTargetGateset`s are possible for a given gateset, a pre-determined default will be
+    applied, but all available `CompilationTargetGateset`s can be accessed via the `target_gatesets`
+    property.
+
+    Users typically interact with this class by accessing its instantiations via
+    `GoogleDevice.metadata.gateset`.
+
+    Examples:
+
+    ```
+    # View what gates and `CompilationTargetGatesets` are supported by the device.
+    print(gateset)
+
+    # Checks whether a gate is valid for the device.
+    cirq.X in gateset
+
+    # Checks whether the circuit only contains valid gates for the device.
+    gateset.validate(circuit)
+
+    # Transform the circuit into one that can be executed on the device.
+    cirq.optimize_for_target_gateset(circuit, gateset=gateset)
+
+    # Choose your own target gateset, if the device supports multiple.
+    cirq.optimize_for_target_gateset(circuit, gateset=gateset.target_gatesets[1])
+    """
+
+    def __init__(
+        self,
+        device_name,
+        *device_gates: Union[Type[cirq.Gate], cirq.Gate, cirq.GateFamily],
+    ) -> None:
+        self._device_name = device_name
+        self._device_gateset = cirq.Gateset(*device_gates)
```

Comment:
It's to distinguish between device gates, which come from the `DeviceSpecification` proto and passed into this gateset as a constructor, and equivalence GateFamilies including `FSimGateFamily` and the future `PhasedXZGateFamily`. This distinction is useful in two places:
* The string representation can explicitly label device gates, and either skip equivalence GateFamilies (i.e. the current implementation), or label them explicitly in the output as well. This distinction clarifies to the user which are the native gate instances / types the device supports, and which are the gate instances / types that are accepted by being equivalent to the native ones.
* Wrapping device gates in a `Gateset` makes `CompilationTargetGateset` and equivalence GateFamily generation a little cleaner.

In both cases, device gates could be *any* collection type and doesn't necessarily need to be a `cirq.Gateset()`, but this representation makes the implementation a bit cleaner.

The base gateset includes both device gates and equivalence GateFamilies, which are both necessary to perform gate validations.

Reply:
I also don't like the pattern of a class inheriting from, and containing instances of, the same base class.  Can we get on a call (maybe the three of us?) to discuss this in more detail?

Reply:
I'm largely on the same page as per our conversation offline, although I do have reservations about keeping a list of target gatesets.

I assume (but perhaps incorrectly) that circuit validation against a Google device gateset is a more common use case than circuit transformation. From our discussion, all target gatesets in the list necessarily have to have the same gateset for validation purposes. It's additional cognitive overhead for the user to have to choose a target gateset if they only want to validate their circuit. I also assume (also perhaps incorrectly) that most users who do want to transform their circuits won't care about which target gateset to use, as long as their circuit can be transformed into a form that's valid for the device, and users would like to be suggested a default that works the best with the device.

These are the reasons why I arrived at a design which
1. Surfaces just a single gateset via `device.metadata` for ease of validation
2. In the case where multiple target gatesets are valid, provides an opinionated default target gateset behavior for this gateset, but still provides a way to access all possible target gatesets.

We could perhaps accomplish the same thing at the metadata level rather than being inside a new wrapper gateset implementation like I have here.

Let me know if my assumptions are correct. For now though, I could implement your proposal and gather some user feedback first.